### PR TITLE
Load files with periods with no data

### PR DESCRIPTION
**Pull request by jclarkeSTFC**
It's possible for a user to set the number of periods in a run independently of how many periods are actually in the data. One case is where the number of periods specified in the file is greater than the number in the data, i.e. there are periods with no data. This fix will now let you load the file in this case, and will give you a warning message about it. If the number of periods specified is less than the number in the data, then the file will still refuse to load.

#### Summary of work

- Added logic for case of nperiods > number of periods in data, including warning message
- Added test to check the problem file in #33729 loads

Fixes #33729. 

### To test:
The file in the referenced issue (`LARMOR00062766.nxs`) should load into one workspace. `LARMOR00003368.nxs` is an example of a file with multiple periods (all with data), it's the one used in several of the existing tests. Other event files should still load okay.

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Comment by jhaigh0** [comment]
Just wondering if you know if its possible for the file to be something like:
```
period 1: has data
period 2: no data
period 3: has data
```
and if this change would work in that case. Or is that a case we don't have to worry about.

File: Framework/DataHandling/src/LoadEventNexus.cpp [hunk 1]
```diff
@@ -574,8 +575,11 @@ LoadEventNexus::runLoadNexusLogs(const std::string &nexusfilename, T localWorksp
     if (run.hasProperty("period_log")) {
       auto *temp = run.getProperty("period_log");
       // Check for corrupted period logs
+      std::string status = "";
       std::unique_ptr<TimeSeriesProperty<int>> tempPeriodLog(dynamic_cast<TimeSeriesProperty<int> *>(temp->clone()));
-      checkForCorruptedPeriods(std::move(tempPeriodLog), periodLog, nPeriods, nexusfilename);
+      nPeriods = checkForCorruptedPeriods(std::move(tempPeriodLog), periodLog, nPeriods, nexusfilename, status);
```
Comment (jhaigh0) [inline Framework/DataHandling/src/LoadEventNexus.cpp:580]: Is `nPeriods` used anywhere after being returned?

File: Framework/DataHandling/src/LoadEventNexus.cpp [hunk 2]
```diff
@@ -740,12 +747,16 @@ std::shared_ptr<BankPulseTimes> LoadEventNexus::runLoadNexusLogs(
  * @param nPeriods :: the value in the nperiods log of the run. Number of
  * expected periods
  * @param nexusfilename :: the filename of the run to load
+ * @param status :: will contain any status message. Empty if no problems.
  */
```
Comment (jhaigh0) [inline Framework/DataHandling/src/LoadEventNexus.cpp:751]: Can you add an `@returns` (or is it `@return`?) to this doc string now that it returns something.

File: Framework/DataHandling/src/LoadEventNexus.cpp [hunk 3]
```diff
@@ -677,9 +681,12 @@ std::shared_ptr<BankPulseTimes> LoadEventNexus::runLoadNexusLogs(
     // Get the period log. Map of DateAndTime to Period int values.
     if (run.hasProperty("period_log")) {
       auto *temp = run.getProperty("period_log");
+      std::string status = "";
       // Check for corrupted period logs
       std::unique_ptr<TimeSeriesProperty<int>> tempPeriodLog(dynamic_cast<TimeSeriesProperty<int> *>(temp->clone()));
-      checkForCorruptedPeriods(std::move(tempPeriodLog), periodLog, nPeriods, nexusfilename);
+      nPeriods = checkForCorruptedPeriods(std::move(tempPeriodLog), periodLog, nPeriods, nexusfilename, status);
+      if (!status.empty())
+        alg.getLogger().warning(status);
```
Comment (jhaigh0) [inline Framework/DataHandling/src/LoadEventNexus.cpp:689]: Why not send the logger warning directly from `checkForCorruptedPeriods`?

**Review by jhaigh0 — CHANGES_REQUESTED** [review]
Solves the issue but had some questions on the implementation

**Comment by jclarkeSTFC** [comment]
> Just wondering if you know if its possible for the file to be something like:
> 
> ```
> period 1: has data
> period 2: no data
> period 3: has data
> ```
> 
> and if this change would work in that case. Or is that a case we don't have to worry about.

I don't know if that can happen or not, but if it did then I think it would work because `nperiods` would be 3, and the detected number of periods would also be 3. I think you'd end up with an empty workspace sandwiched between workspaces 1 and 2.

File: Framework/DataHandling/src/LoadEventNexus.cpp (hunk 2 above)
Comment (jclarkeSTFC) [reply Framework/DataHandling/src/LoadEventNexus.cpp:751]: Yes I'll do that

File: Framework/DataHandling/src/LoadEventNexus.cpp (hunk 3 above)
Comment (jclarkeSTFC) [reply Framework/DataHandling/src/LoadEventNexus.cpp:689]: `checkForCorruptedPeriods` is static so it doesn't know about the logger, unless it got passed as an argument, could do that instead.

File: Framework/DataHandling/src/LoadEventNexus.cpp (hunk 1 above)
Comment (jclarkeSTFC) [reply Framework/DataHandling/src/LoadEventNexus.cpp:580]: `nPeriods` is an output from the parent method, it's used [here](https://github.com/mantidproject/mantid/blob/7680bac7757ce51ac42e7fd3fb6b2ad8829f1036/Framework/DataHandling/src/LoadEventNexus.cpp#L918C42-L918C42).

File: Framework/DataHandling/src/LoadEventNexus.cpp (hunk 3 above)
Comment (jhaigh0) [reply Framework/DataHandling/src/LoadEventNexus.cpp:689]: ah I see, no problem to leave it as is

File: Framework/DataHandling/src/LoadEventNexus.cpp (hunk 1 above)
Comment (jhaigh0) [reply Framework/DataHandling/src/LoadEventNexus.cpp:580]: ah I see thanks

**Review by jhaigh0 — APPROVED** [review]

**Pull request merged**